lib/agent.js

```diff
@@ -104,9 +104,7 @@ class Agent extends Dispatcher {
         throw new InvalidArgumentError('opts must be a object.')
       }
 
-      if (typeof opts.origin !== 'string' || opts.origin === '') {
-        throw new InvalidArgumentError('opts.origin must be a non-empty string.')
-      }
+      const { origin } = util.parseOrigin(opts.origin)
```

Comment:
I'm a little concerned about the performance implications. Is it negligable?

Reply:
Parsing URLs is _slow_, we should not be creating a fresh `URL` instance every time if possible.

Reply:
origin is just used as a key here so we won't actually need to validate it.

Reply:
```
let origin
if (opts.origin && typeof opts.origin === 'string') {
  origin = opts.origin
} else if (opts.origin instanceof URL) {
  origin = opts.origin.origin
} else {
  throw new InvalidArgumentError('opts.origin must be a non-empty string or URL.')
}
```